src/compiler/moduleNameResolver.ts

```diff
@@ -1511,7 +1708,231 @@ namespace ts {
+                }
+                const finalPath = getNormalizedAbsolutePath(pattern ? resolvedTarget.replace(/\*/g, subpath) : resolvedTarget + subpath, state.host.getCurrentDirectory?.());
+
+                return toSearchResult(withPackageId(scope, loadJSOrExactTSFileName(extensions, finalPath, /*onlyRecordFailures*/ false, state)));
+            }
+            else if (typeof target === "object" && target !== null) { // eslint-disable-line no-null/no-null
+                if (!Array.isArray(target)) {
+                    for (const key of getOwnKeys(target as MapLike<unknown>)) {
+                        if (key === "default" || state.conditions.indexOf(key) >= 0 || isApplicableVersionedTypesKey(state.conditions, key)) {
+                            const subTarget = (target as MapLike<unknown>)[key];
+                            const result = loadModuleFromTargetImportOrExport(subTarget, subpath, pattern);
+                            if (result) {
+                                return result;
+                            }
+                        }
+                    }
+                    return undefined;
+                }
+                else {
+                    if (!length(target)) {
+                        if (state.traceEnabled) {
+                            trace(state.host, Diagnostics.package_json_scope_0_has_invalid_type_for_target_of_specifier_1, scope.packageDirectory, moduleName);
+                        }
+                        return toSearchResult(/*value*/ undefined);
+                    }
+                    for (const elem of target) {
+                        const result = loadModuleFromTargetImportOrExport(elem, subpath, pattern);
+                        if (result) {
+                            return result;
+                        }
+                    }
+                }
+            }
+            else if (target === null) { // eslint-disable-line no-null/no-null
+                if (state.traceEnabled) {
+                    trace(state.host, Diagnostics.package_json_scope_0_explicitly_maps_specifier_1_to_null, scope.packageDirectory, moduleName);
+                }
+                return toSearchResult(/*value*/ undefined);
+            }
+            if (state.traceEnabled) {
+                trace(state.host, Diagnostics.package_json_scope_0_has_invalid_type_for_target_of_specifier_1, scope.packageDirectory, moduleName);
+            }
+            return toSearchResult(/*value*/ undefined);
+        }
+
+        function isApplicableVersionedTypesKey(conditions: string[], key: string) {
+            if (conditions.indexOf("types") === -1) return false; // only apply versioned types conditions if the types condition is applied
+            if (!startsWith(key, "types@")) return false;
+            const range = VersionRange.tryParse(key.substring("types@".length));
+            if (!range) return false;
+            return range.test(version);

```

Comment:
where is version declared?

Reply:
In `core.ts`. I'd normally write `ts.version` to clarify, but our lint settings don't like that, so... 🤷